[FEATURE] Création de la méthode d'authentification utilisant spécifiquement un mot de passe (PIX-1365).

chrc opened the pull request:
## :unicorn: Problème
Suite à la création de la table authentication-methods, les différentes méthodes de connexion existantes doivent être migrées.
Le but de ce ticket est d'effectuer la migration des données et des fonctions relatives à la méthode de connexion PIX (avec un mot de passe).

## :robot: Solution
- [x] Supprimer les champs `password` et `shouldChangePassword` du model `User`
- [x] Utilisation de la table `authentication-methods`
  * Dans le champ `authenticationComplement` de type json, écrire le `password` et `shouldChangePassword`
- [x] Mettre à jour les databaseBuilders concernés
- [x] Mettre à jour les domainBuilders concernés
- [x] Mettre à jour les méthodes des repositories manipulant l'attribut `user.password`
- [x] Créer le user-service
- [x] Mettre à jour les méthodes des usecases manipulant l'attribut `user.password`
- [x] Mettre à jour les seeds concernés
- [x] Mettre à jour dans Cypress, les fixtures `users`
- [x] Créer dans Cypress, les fixtures `authentication-methods`
- [x] Créer dans Cypress, le scénario Pix Orga `Je veux créer un identifiant pour un élève ayant la méthode de connexion Mediacentre`
- [ ] Mettre à jour les tests Mirage utilisant `password` et `shouldChangePassword`

## :rainbow: Remarques
* Temporairement, on garde les colonnes `password` et `shouldChangePassword` dans la table `users`.
* Ils seront supprimés dans un prochain ticket (lancement d'un script SQL).

## :100: Pour tester
Recette générale de tous les fronts : 
* Authentification
* Création user
* Modification user/password/etc.
* Lecture user
* Méthodes de connexion

[Ici](https://github.com/1024pix/pix/pull/2220#pullrequestreview-552458204) la liste plus exhaustive des tests de non regression, établie par @AndreiaPena

## :warning:  *MISE EN PRODUCTION*  :warning:
Juste après la MEP, deux scripts SQL doivent être exécutés (cf. JIRA [PIX-1365](https://1024pix.atlassian.net/browse/PIX-1365)).

chrc committed:
[API] Add users password infos migration script

chrc committed:
[API] BSR - AlreadyExistingEntity name to Error

chrc committed:
[API] BSR - Sort list of errors

chrc committed:
[API] Add AuthenticationMethodNotFoundError

chrc committed:
[API] Add database-builder methods

Add buildAuthenticationMethod.buildWithHashedPassword method
Add private buildPasswordAuthenticationMethod method into buildUser()

chrc committed:
[API] Add authenticationMethods to user model

chrc committed:
[API] BSR test-helper

chrc committed:
[API] Create password validator

chrc committed:
[API] Update user validator

Extract password validation
Call extracted function

chrc committed:
[API] Add authenticationMethods User Bookshelf

chrc committed:
[API] Update user repository methods

BSR on required files
Add domainTransaction to create()
Delete createAndReconcileUserToSchoolingRegistration()
Delete updatePasswordThatShouldBeChanged()
Delete updateExpiredPassword()
Rename getByUsernameOrEmailWithRoles() to getByUsernameOrEmailWithRolesAndPassword()
Rename updateUsernameAndPassword() to updateUsername()

chrc committed:
[API] Update authenticationMethod repository methods

BSR on required files
Add updateExpiredPassword()
Add updateOnlyShoulChangePassword()
Add updatePassword()
Add updatePasswordThatShouldBeChanged()

chrc committed:
[API] Update schoolingRegistration repo methods

BSR on required files
Add updateUserIdWhereNull()

chrc committed:
[API] BSR into repositories integration tests

Delete authentication-methods table after each tests
Delete password from user attributes

chrc committed:
[API] Rename encryption service method

Rename check() to checkPassword()
Use Destructuring

chrc committed:
[API] Update authentication service

chrc committed:
[API] Create user service

chrc committed:
[API] Update create user usecases

Use injection for authenticationMethodRepository and userService
BSR tests create-and-reconcile-user-to-schooling-registration
BSR tests create-user

chrc committed:
[API] Update generate-username-. usecase

Use injection for authenticationMethodRepository and userService
BSR tests generate-username-with-temporary-password

chrc committed:
[API] Update update-* usecases

Use injection for authenticationService, encryptionService and authenticationMethodRepository
BSR tests update-expired-password
BSR tests update-schooling-registration-dependent-user-password
BSR tests update-user-password

chrc committed:
[API] Update authenticateUser usecase

BSR tests authenticate-user

chrc committed:
[API] Update user controller

BSR tests user-controller
BSR tests users-controller-update-password

chrc committed:
[API] Update Schooling-registration-. controller

BSR tests schooling-registration-dependent-user controller

chrc committed:
[API] Update controllers tests with authentication

Delete authentication-methods table data after each tests
BSR test password-controller

chrc committed:
[API] BSR acceptance tests

chrc committed:
[API] Update AuthenticationMethod model

chrc committed:
[API] Update createUserAndReconcile.. usecase

BSR tests createUserAndReconcileToSchoolingRegistrationFromExternalUser

pix-service commented:
I'm deploying this PR to these urls:

- App: https://app-pr2220.review.pix.fr
- Orga: https://orga-pr2220.review.pix.fr
- Certif: https://certif-pr2220.review.pix.fr
- Admin: https://admin-pr2220.review.pix.fr
- API: https://api-pr2220.review.pix.fr/api/

 Please check it out!

chrc committed:
[API] Update organizations SCO buider seed

chrc committed:
[E2E] Update users & authentication-methods fixtures

MathieuGilet commented on api/lib/infrastructure/repositories/authentication-method-repository.js:
```diff
@@ -18,18 +25,95 @@ function _toDomainEntity(bookshelfAuthenticationMethod) {
+      });
+  },
+
+  updateExpiredPassword({ userId, hashedPassword }) {
```
En reparler en recap tech.

octo-topi commented:
**Remarque :** A props du commit "[API] BSR - AlreadyExistingEntity name to Error"
Tu peux utiliser le linter pour expliciter une convention de nommage https://eslint.org/docs/rules/id-match

MathieuGilet commented on api/db/migrations/20201113105430_migrate-password-and-shouldChangePassword-to-authentication-methods.js:
```diff
@@ -0,0 +1,48 @@
+
+  const usersWithPasswordInformations = await knex.raw(sqlRequest);
+
+  return batch(knex, usersWithPasswordInformations.rows, (user) => {
```
**Problème**: l'instruction batch ici ne gère que des lots de 10 éléments ce qui va rendre le script beaucoup trop long à exécuter. Le plus simple serait d'utiliser une requête INSERT INTO SELECT qui va se charger d'effectuer toutes les insertions en une seule fois:
`INSERT INTO "authentication-methods"("identityProvider", "userId", "authenticationComplement") SELECT 'PIX', id, jsonb_build_object('password', "password", 'shouldChangePassword', "shouldChangePassword") FROM "users" WHERE ("password" = '') IS FALSE;`

chrc replied:
Done

MathieuGilet commented on api/lib/domain/models/AuthenticationMethod.js:
```diff
@@ -47,7 +46,7 @@ const validationSchema = Joi.object({
     { is: identityProviders.PIX, then: Joi.object().instance(PasswordAuthenticationMethod).required() },
     { is: identityProviders.POLE_EMPLOI, then: Joi.object().instance(PoleEmploiAuthenticationComplement).required() },
-    { is: identityProviders.GAR, then: Joi.any().forbidden() },
+    { is: identityProviders.GAR, then: Joi.any().empty() },
```
**Question**: Pourquoi ce changement ? La doc sur forbidden dit: "Marks a key as forbidden which will not allow any value except undefined. Used to explicitly forbid keys.". Il me semble que l'on veut justement que la valeur d'`authenticationComplement` pour le provider GAR soit uniquement `undefined`

chrc replied:
A mon sens l'attribut `authenticationComplement` fait partie du model.
Il peut être `null` ou `undefined`, mais il doit exister.

MathieuGilet commented on api/db/database-builder/factory/build-authentication-method.js:
```diff
@@ -47,7 +47,38 @@ buildAuthenticationMethod.buildPasswordAuthenticationMethod = function({
+  });
+};
+
+buildAuthenticationMethod.buildWithHashedPassword = function({
```
**Question**: Pourquoi ajouter cette fonction ? Il me semble qu'elle est similaire à celle du dessus, dans les deux fonctions le password passé en paramètre est encrypté.

chrc replied:
Done

AndreiaPena reviewed (changes requested):
**Bloquant** Nous avons détecté un souci coté Pix Orga : Impossible d'ajouter une méthode de connexion identifiant à un user en méthode unique Médiacentre.
Message d'erreur : `Authentication method PIX for User ID 100040 not found.`

Problème reglé, il a été testé juste après, ça fonctionne.

chrc committed:
[API] Update authentication-poleEmploiUser usecase

MathieuGilet commented on api/lib/domain/usecases/create-user.js:
```diff
@@ -47,22 +71,36 @@ async function _validateData(user, reCaptchaToken, userRepository, userValidator
+  const isValid = await _validateData({
+    user, password, reCaptchaToken,
+    userRepository,
+    userValidator, reCaptchaValidator,
```
**BSR**: retirer le userValidator en paramètre. Ca aurait du être fait dans un refacto de mai 2018 donc ce n'est pas "urgent".

chrc replied:
Done

MathieuGilet commented on api/lib/domain/services/user-service.js:
```diff
@@ -0,0 +1,125 @@
+  });
+}
+
+async function createUserWithPassword({
```
**Question**: pourquoi exporter ce bout de code dans un service alors qu'il n'est utilisé que dans le usecase `createUser` ?

chrc replied:
Pour exporter, refactorer et tester cette fonction, qui fait appel à deux repositories distincts.

MathieuGilet commented on api/lib/domain/services/user-service.js:
```diff
@@ -0,0 +1,125 @@
+
+  return {
+    savedUser,
+    savedAuthenticationMethod,
```
**Remarque**: je n'ai pas l'impression que `savedAuthenticationMethod` soit utilisé dans le usecase et le controller, tu pourrais donc ne pas le retourner.

chrc replied:
Done

MathieuGilet commented on api/lib/infrastructure/serializers/jsonapi/user-serializer.js:
```diff
@@ -16,6 +16,7 @@ module.exports = {
         'memberships', 'certificationCenterMemberships',
         'pixScore', 'scorecards', 'profile',
         'campaignParticipations', 'hasSeenAssessmentInstructions', 'isCertifiable',
+        'authenticationMethods',
```
**Question**: A quel moment tu penses que l'on aura besoin des méthodes d'authentification dans le/les front(s) ? De plus, si on en a besoin et sans création du model `authenticationMethod` dans le/les front(s) ça ne devrait pas fonctionner.

chrc replied:
On en aura besoin au moins lorsqu'on voudra savoir si le user doit changer de mot de passe (`shouldChangePassword`).
Et en effet, il faut faire les mises à jour côté front, pour lire cette information dans le model `user`

chrc replied:
Vu avec @octo-thlo et @MathieuGilet, le `user-serializer` ne "remontera" pas `authenticationMethod`.

AndreiaPena reviewed (commented):
**Tests réalisés sur les RA avec Thierry :** 

Pix App

- Création de compte ( double mire : identifiant ) => OK
- Création de compte ( double mire : adresse e-mail ) => OK
- Création de compte ( classique ) => OK
_(  password "" et shouldChangePassword à false pour **users** => check 
ajout d'une ligne dans **authenticationMethods** avec le password rempli et shouldChangePassword à false => check
shouldChangePassword reste à false **(authenticationMethods)** pendant la phase de Réinitialisation de MDP => check  
 )_
- Connexion MDP => OK
- Connexion via double mire => OK
- Connexion avec utilisateur ayant mail + identifiant => OK
- Reset MDP avec utilisateur ayant mail + identifiant => OK
- Reset mot de passe => OK
- Affichage du mail page de profil => OK

Pix Orga

- Connexion  => OK
 - Multi-accès => OK
-  Invitations => OK
- Création de compte après reception d'une invitation => OK
 _(  ajout d'une ligne dans **users** avec password "" et shouldChangePassword à false  => check 
ajout d'une ligne dans **authenticationMethods** avec le password rempli et shouldChangePassword à false => check
 )_
- Page élèves : 
- Ajout méthode de connexion identifiant lorsque Médiacentre unique => OK
_(  password "" et shouldChangePassword à false pour **users** avant et après ajout Identifiant et après réinitialisation de MDP => check 
ajout d'une ligne dans **authenticationMethods** avec le password rempli et shouldChangePassword à true => check
shouldChangePassword passe à false **(authenticationMethods)** après réinitialisation du MDP => check 
 )_

- Ajout méthode de connexion identifiant lorsque Adresse-mail unique => OK
_(  password "" et shouldChangePassword à false pour **users** avant et après ajout Identifiant => check 
ligne inchangée dans **authenticationMethods** avec le password rempli et shouldChangePassword à false => check
shouldChangePassword passe à true **(authenticationMethods)** lorsqu'on génère un MDP à usage unique => check 
repasse à false après réinitialisation de MDP et ligne dans **users** inchangée pour password et shouldChangePassword)_

Pix Certif

- Connexion => OK

Pix Admin

- Connexion (vérifier les droits Pix Master) => OK
- Ajout d’un membre => OK
- Modif d’un e-mail utilisateur => OK
- Anonymisation => OK
- Dissociation => OK
- Affichage des infos utilisateurs => OK

AndreiaPena commented on api/lib/domain/errors.js:
```diff
@@ -662,26 +668,24 @@ class NotImplementedError extends Error {
   AssessmentNotCompletedError,
   AssessmentResultNotCreatedError,
-  ArchivedCampaignError,
+  AuthenticationMethodNotFoundError,
```
**Non-bloquant suggestion** : Rajout d'un test unitaire sur errors_tests.js
 ```suggestion
 it('should export a AuthenticationMethodNotFoundError', () => {
    expect(errors.AuthenticationMethodNotFoundError).to.exist;
  });
```

chrc replied:
Done

chrc committed:
[ORGA] Add id attribute to SelectInput component

We will use this into E2E Pix Orga test

chrc committed:
[E2E] Add Pix Orga Username generation scenario

Generate an username and password, for a student who has only the Mediacentre connexion type.

MathieuGilet commented on api/lib/infrastructure/repositories/user-repository.js:
```diff
@@ -125,6 +170,7 @@ function _toDomain(userBookshelf) {
     certificationCenterMemberships: _toCertificationCenterMembershipsDomain(userBookshelf.related('certificationCenterMemberships')),
     pixRoles: _toPixRolesDomain(userBookshelf.related('pixRoles')),
     hasSeenAssessmentInstructions: Boolean(userBookshelf.get('hasSeenAssessmentInstructions')),
+    authenticationMethods: _toAuthenticationMethodsDomain(userBookshelf.related('authenticationMethods')),
```
**Suggestion**: J'ai continué à réfléchir sur l'ajout de l'attribut `authenticationMethods` ici et l'impact inévitable sur le serializer et donc le/les fronts qui utilisent ces données. Je pense qu'il aurait été plus simple de conserver les attributs `password` et `shoudlChangePassword` sur le model `User` et de les renseigner depuis la table authentication-methods. Ca réduit, je pense, les changements apportés au code en limitant les changements au répo voire aux usecases. Si tu penses que ce n'est pas la bonne solution, il faut au moins suppimer password et shouldChangePassword du mofel User afin d'éviter les endroits où on pourrait encore lire dans ces colonnes par erreur vu qu'on ne les supprime pas encore.

chrc replied:
Dans le domain model `User`
* `password` et `shouldChangePassword` ont été supprimés
* `authenticationMethods` a été ajouté

Après plusieurs analyses, j'ai voulu garder la logique prise pour `GAR` et `Pole emploi` : déplacer complètement les attributs et type de connexion vers un model distinct `AuthenticationMethod`.
J'ai fait le choix de modifier au minimum les `usecases`, mais de travailler plus sur les interfaces/fonctions des `repositories` et `services` concernés.

chrc replied:
Voir plus haut.

jonathanperret commented on api/db/migrations/20201113105430_migrate-password-and-shouldChangePassword-to-authentication-methods.js:
```diff
@@ -0,0 +1,34 @@
+};
+
+exports.down = async (knex) => {
+  const pixAuthenticationMethods = await knex(TABLE_AUTHENTICATION_METHODS)
+    .select('userId', 'authenticationComplement')
+    .where({ identityProvider: AuthenticationMethod.identityProviders.PIX });
+
+  await batch(knex, pixAuthenticationMethods, (authenticationMethod) => {
+    return knex(TABLE_USERS)
+      .where({ id: authenticationMethod.userId })
+      .update({
+        password: authenticationMethod.authenticationComplement.password,
+        shouldChangePassword: authenticationMethod.authenticationComplement.shouldChangePassword,
+      });
+  });
+
+  return knex(TABLE_AUTHENTICATION_METHODS)
+    .where({ identityProvider: AuthenticationMethod.identityProviders.PIX })
+    .delete();
```
**question** : pourquoi est-ce que la migration `down` n'est pas faite côté PG comme le `up` ?
Vu le nombre d'utilisateurs en production j'ai peur que l'approche consistant à charger en mémoire les données d'authentification de tous les utilisateurs ne passe pas 🤔

chrc replied:
Done

AndreiaPena commented on api/tests/unit/domain/validators/user-validator_test.js:
```diff
@@ -195,29 +160,24 @@ describe('Unit | Domain | Validators | user-validator', function() {
         });
 
-        it('should reject with errors on firstName, lastName, email and password when firstName, lastName, email and password have a maximum length of 255', async () => {
+        it('should reject with errors on firstName, lastName, email and password when firstName, lastName and email have a maximum length of 255', async () => {
```
**Chipotage/Remarque** : Oubli d'un password dans l'intitulé du test.

chrc replied:
Done

thloesch commented on api/db/database-builder/factory/build-user.js:
```diff
@@ -1,40 +1,75 @@
 
 const PIX_MASTER_ROLE_ID = 1;
 
+function _buildPasswordAuthenticationMethod({
```
**Remarque non-bloquante** : Avec @chrc et @MathieuGilet on a discuté d'un refacto pour ne plus avoir à utiliser cette méthode privée.

Nous avons créé un ticket [ici](https://1024pix.atlassian.net/browse/PIX-1812)

octo-topi commented on api/db/database-builder/factory/build-authentication-method.js:
```diff
@@ -32,22 +32,24 @@ const buildAuthenticationMethod = function({
 } = {}) {
 
-  password = isUndefined(password) ? encrypt.hashPasswordSync(faker.internet.password()) : encrypt.hashPasswordSync(password);
+  const password = isUndefined(hashedPassword) ? encrypt.hashPasswordSync(faker.internet.password()) : hashedPassword;
```
Remarque 
Une PR récente a supprimé des password aléatoires, peut-être est-ce applicable ici aussi ?
https://github.com/1024pix/pix/blob/dev/api/db/database-builder/factory/build-user.js#L19

chrc replied:
Quel était le problème ?

octo-topi replied:
Le problème de l'autre PR et de celle-ci était de ne pas pouvoir se connecter 
Plutôt que de tester si le password est renseigné, tu mets la valeur par défaut `pix123` sur le paramètre
Du coup
- tu connais le mot de passe pour te connecter
- le code est réduit à un appel de fonction
```suggestion
  const hashedPassword = encrypt.hashPasswordSync(password);
```

chrc replied:
Done

octo-topi commented on api/tests/test-helper.js:
```diff
@@ -38,13 +35,12 @@ function generateIdTokenForExternalUser(externalUser) {
     lastName: 'Papa',
     email: 'super.papa@example.net',
-    password: 'abcd1234',
+    rawPassword: 'abcd1234',
```
Chipotage
Je cherchais l'équivalent de "non chiffré"
J'ai pas trouvé mieux que `unhashed` , même dans [cette RFC](https://tools.ietf.org/id/draft-farrell-decade-ni-03.html)
```suggestion
    unhashedPassword: 'abcd1234',
```

jonathanperret replied:
@octo-topi je ne trouve pas le mot "unhashed" dans [la RFC en question] ?

Le mot `unhashed` à mon sens fait plutôt référence au résultat d'une attaque sur un hash, où on réussit à retrouver un mot de passe à partir d'un hash.

Le problème ici je trouve que ce n'est pas d'appeler `password` le mot de passe — après tout c'est parfaitement correct — mais plutôt d'appeler `password` le résultat de la fonction de hachage alors que tout le but est que ce ne soit précisément _pas_ un mot de passe (au sens de quelque chose qui doit absolument rester secret). Si on utilisait par exemple `passwordHash` là où on utilise `password` on n'aurait pas besoin d'ajouter `raw-`, `unhashed-`, etc.

octo-topi replied:
J'ai corrigé mon commentaire qui prêtait à confusion: même la RFC ne définit pas de terrme.
Je dis :100: à `hashedPassword` ou `hashFromPassword` plutôt que de renommer la source `plainPassword`

jonathanperret replied:
Si on pinaille, je continue de préférer `passwordHash` : c'est _un hash_ du _mot de passe_, deux objets de nature très différente. `hashedPassword` m'évoque un objet qu'on aurait modifié sans changer sa nature. Par exemple si j'ai la liste `[1,2,3]` alors je trouve ça plus parlant de dire que `6` est la _somme de la liste_ plutôt que _la liste sommée_ — alors que `[1,2]` serait _la liste tronquée_…

chrc replied:
Done

octo-topi replied:
Implémenté dans https://github.com/1024pix/pix/pull/2310

octo-topi commented on api/tests/tooling/domain-builder/factory/build-authentication-method.js:
```diff
@@ -14,35 +25,63 @@ const buildAuthenticationMethod = function({
 };
 
-buildAuthenticationMethod.buildPasswordAuthenticationMethod = function({
+buildAuthenticationMethod.buildWithRawPassword = function({
```
Remarque :
Je ne me ferais jamais à cette syntaxe @MathieuGilet  :exploding_head: 
Une fonction qui .. contient des fonctions.. mais qui n'est pas une instance de classe :dizzy_face: 
```` js
node
> const myFunction = function(){ console.log('foo'); };
> myFunction.bar = function (){ console.log('bar'); };
> myFunction();
foo
> myFunction.bar();;
bar
````

MathieuGilet replied:
Vive le JS ;)

jonathanperret replied:
@octo-topi pas une instance ?

```
$ node
> (function(){}) instanceof Function
true
```

😉

octo-topi replied:
![image](https://user-images.githubusercontent.com/56302715/102873592-9cd99400-4441-11eb-877e-6d76f27dee84.png)
Bertrand Meyer le répète au moins..... tout ça de fois !

octo-topi commented on api/tests/tooling/domain-builder/factory/build-user.js:
```diff
@@ -1,16 +1,19 @@
-    email = 'jean.bono@example.net',
-    username = 'jean.bono1234',
-    password = 'liuehrfi128743KUUKNSUkuz12Ukun',
+    id = faker.random.number(),
```
Remarque
Y'en a qui ont essayé, ils ont eu des problèmes :sweat_smile: 
Je me demande s'il n'y a pas risque de collision/flaky test. 
Pourquoi ne pas faire une séquence à la mano ?

```
const createSequence = function () {
    let count = 0;
    return function (){ return count++; }
};

const next = createSequence();
console.log(next());
> 0
console.log(next());
> 1
```

octo-topi commented on api/lib/domain/validators/password-validator.js:
```diff
@@ -0,0 +1,30 @@
+const Joi = require('@hapi/joi');
+const XRegExp = require('xregexp');
+
+const { passwordValidationPattern } = require('../../config').account;
```
Chouette
:bow:  pour l'encapsulation

Chipotage
Si tu nomme le pattern de refacto que tu utilises comme Fowler dans les commits , ça donne les commits suivants (plutôt que l'énigmatique `Update User validator`)
```
[Extract function] Extract password validation
[Extract function] Call extracted function
```

chrc replied:
Done

MathieuGilet commented on api/lib/domain/services/encryption-service.js:
```diff
@@ -21,9 +21,9 @@ module.exports = {
   },
 
-  check: (plain, hash) => {
+  checkPassword: ({ rawPassword, hashedPassword }) => {
```
**Remarque**: J'aurais laissé le nommage actuel personnellement même si on n'encrypte que des mots de passe pour le moment.

AndreiaPena commented on api/tests/tooling/domain-builder/factory/build-authentication-method.js:
```diff
@@ -1,10 +1,21 @@
+const User = require('../../../../lib/domain/models/User');
 const AuthenticationMethod = require('../../../../lib/domain/models/AuthenticationMethod');
 
+function _buildUser() {
```
**A discuter en recap tech**

octo-topi commented on api/lib/infrastructure/data/user.js:
```diff
@@ -1,6 +1,7 @@
 const User = require('../../domain/models/User');
 const PixRole = require('../../domain/models/PixRole');
+
+const Bookshelf = require('../bookshelf');
```
Chipotage
Pas fan de cette BSR: l'intention b'est pas claire et le nom du commit ne m'a pas aidé
`[API] Add authenticationMethods User Bookshelf`

Je m'attends à un tri:
- soit alphabétique
- soit package (dépendances npm)  / fichier de l'app
- soit composant qui teste / composant testé

Là c'est un groupement.. ? couche clean archi :thinking:

octo-topi commented on api/lib/infrastructure/repositories/user-repository.js:
```diff
@@ -143,11 +189,18 @@ function _setSearchFiltersForQueryBuilder(filter, qb) {
 
 function _adaptModelToDb(user) {
-  return omit(user, [
+  const changedUser = omit(user, [
```
Chipotage
```suggestion
  const trimmedUser = omit(user, [
```

```suggestion
  const userToBeSaved = omit(user, [
```

```suggestion
  const userReadyToBePersisted = omit(user, [
```

chrc replied:
Done

octo-topi commented on api/lib/infrastructure/repositories/user-repository.js:
```diff
@@ -458,14 +483,25 @@ module.exports = {
 
-  updateUsernameAndPassword(id, username, hashedPassword) {
+  updateUsername({
+    userId,
```
Chipotage
```suggestion
    id,
```
Voiture.couleurVoiture
Il n'y a pas d'ambiguïté ici, non ?

chrc replied:
Done

octo-topi commented on api/lib/infrastructure/data/user.js:
```diff
@@ -1,6 +1,7 @@
 const User = require('../../domain/models/User');
 const PixRole = require('../../domain/models/PixRole');
+
+const Bookshelf = require('../bookshelf');
```
Remarque
Si seulement .. 

```suggestion
const BookshelfValidator = require('../bookshelf');
```
ou
```suggestion
const Bookshelf = require('../bookshelf-validator');
```
ou
```suggestion
const Bookshelf = require('../validator/bookshelf');
```

Je verrais qu'on ne parle pas d'un package npm

chrc replied:
Je n'ai pas compris. Quel est le rapport avec validator ?

octo-topi commented on api/lib/infrastructure/repositories/schooling-registration-repository.js:
```diff
@@ -1,21 +1,30 @@
 const _ = require('lodash');
 const bluebird = require('bluebird');
-const { knex } = require('../bookshelf');
```
:rofl:

AndreiaPena commented on api/tests/unit/domain/validators/user-validator_test.js:
```diff
@@ -182,7 +147,7 @@ describe('Unit | Domain | Validators | user-validator', function() {
         });
 
-        it('should reject with error when neither email nor username are defined', async() => {
+        it('should reject with error when neither email nor username are defined', async () => {
```
**Non-Bloquant - Question** Lorsqu'on crée un utilisateur qui vient du GAR (ENT) il n'a ni email, ni username.
Pourquoi ce test de validation est important ?

MathieuGilet replied:
On utilise pas le user-validator lors de la création d'un utilisateur provenant du GAR mais lors de la création d'un utilisateur "classique" et dans ce cas on veut s'assurer qu'il possède soit un email soit un identifiant.

octo-topi commented on api/lib/domain/services/encryption-service.js:
```diff
@@ -21,9 +21,9 @@ module.exports = {
   },
 
-  check: (plain, hash) => {
+  checkPassword: ({ rawPassword, hashedPassword }) => {
```
Remarque
Est-ce que le besoin derrière renommer check en checkPassword est de comprendre l'intention ? 
Si oui, je propose de
- renommer `encryption-service` en `password-service` (vu que 
- `checkPassword ` en `passwordMatch` ou `isPasswordMatching` ou `isPasswordValid`

chrc replied:
C'était pour appliquer le pattern de nommage des autres fonctions.

octo-topi commented on api/tests/unit/domain/services/encryption-service_test.js:
```diff
@@ -1,47 +1,45 @@
-      const encryptedPassword = bcrypt.hashSync('my-real-password', 1);
-      const password = 'my-expected-password';
+      const rawPassword = 'my-expected-password';
+      const hashedPPassword = 'ABCDEF1234';
```
Chipotage
```suggestion
      const hashedPassword = 'ABCDEF1234';
```

chrc replied:
Done

octo-topi commented on api/tests/unit/domain/services/encryption-service_test.js:
```diff
@@ -1,47 +1,45 @@
     });
 
-    it('should resolve when passwords are matching', () => {
+    it('should resolve when passwords are matching', async () => {
```
Chipotage
```suggestion
    it('should resolve to undefined when passwords are matching', async () => {
```

chrc replied:
Done

octo-topi commented on api/tests/acceptance/application/users/users-controller-save_test.js:
```diff
@@ -42,60 +59,72 @@ describe('Acceptance | Controller | users-controller', () => {
+
         // when
-        const promise = server.inject(options);
+        const response = await server.inject(options);
```
Chouette
Tellement mieux !

octo-topi commented:
Méta-remarque 
Je mets toujours les BSR/refacto en début de PR 
https://blog.octo.com/en/seven-shades-of-git/

![image](https://user-images.githubusercontent.com/56302715/102639225-019ea100-4159-11eb-957f-794dd933a42b.png)

octo-topi commented:
Hello. Comme vu avec @octo-thlo , j'ai revu le code, pas testé le fonctionnel
 (pour ne pas tester N fois des scénarios nombreux suite aux retours de PR)

AndreiaPena commented:
> **Bloquant** Nous avons détecté un souci coté Pix Orga : Impossible d'ajouter une méthode de connexion identifiant à un user en méthode unique Médiacentre.
> Message d'erreur : `Authentication method PIX for User ID 100040 not found.`
> 
> Problème reglé, il a été testé juste après, ça fonctionne.

thloesch commented on api/tests/integration/infrastructure/repositories/user-repository_test.js:
```diff
@@ -277,35 +288,51 @@ describe('Integration | Infrastructure | Repository | UserRepository', () => {
+        expect(foundUser.cgu).to.equal(expectedUser.cgu);
+      });
+
+      it('should return authenticationMethods associated to the user', async () => {
+        // given
+        const omittedAttributes = ['createdAt', 'updatedAt'];
+        const expectedAuthenticationMethod = omit(passwordAuthenticationMethodInDB, omittedAttributes);
+
+        // when
+        const foundUser = await userRepository.getByUsernameOrEmailWithRolesAndPassword(userInDB.email);
+
+        // then
+        expect(foundUser.authenticationMethods).to.be.an('array');
+        expect(foundUser.authenticationMethods).to.have.lengthOf(1);
+
+        const firstAuthenticationMethod = foundUser.authenticationMethods[0];
+        expect(omit(firstAuthenticationMethod, omittedAttributes))
+          .to.deep.equal(expectedAuthenticationMethod);
```
**Suggestion non-bloquante** : Avec @AndreiaPena on a trouvé la lecture de ce test un peu délicate, notamment à cause des omit. Si on se limite à tester les champs qui nous intéressent (et donc ignorer les `created_at` et `updated_at`), on pourrait se retrouver avec : 

```suggestion
      it('should return authenticationMethods associated to the user', async () => {
        // when
        const foundUser = await userRepository.getByUsernameOrEmailWithRolesAndPassword(userInDB.email);

        // then
        expect(foundUser.authenticationMethods).to.be.an('array');
        expect(foundUser.authenticationMethods).to.have.lengthOf(1);

        const firstAuthenticationMethod = foundUser.authenticationMethods[0];
        expect(firstAuthenticationMethod.identityProvider).to.equal(passwordAuthenticationMethodInDB.identityProvider);
        expect(firstAuthenticationMethod.userId).to.equal(passwordAuthenticationMethodInDB.userId);
                     expect(firstAuthenticationMethod.authenticationComplement).to.deep.equal(passwordAuthenticationMethodInDB.authenticationComplement);
      });
```

chrc replied:
Je dois comparer également le contenu de `authenticationComplement`.

chrc replied:
Done

thloesch commented on api/tests/integration/infrastructure/repositories/user-repository_test.js:
```diff
@@ -776,98 +799,6 @@ describe('Integration | Infrastructure | Repository | UserRepository', () => {
-    });
-  });
-
-  describe('#updateEmail', () => {
```
**Remarque Bloquante** : Pourquoi avoir supprimé ce test ? Nous ne l'avons pas trouvé en cherchant s'il avait été déplacé ailleurs

chrc replied:
Bien vue ! Une mauvaise manip :(

chrc replied:
Done

thloesch commented on api/lib/infrastructure/repositories/user-repository.js:
```diff
@@ -109,6 +114,46 @@ function _toPixRolesDomain(pixRolesBookshelf) {
   });
 }
 
+function _getAuthenticationComplementAndExternalIdentifier(authenticationMethodBookshelf) {
```
**Chipotage non-bloquant** : Je suis plus à l'aise en lisant les méthodes les plus importantes en haut, puis les privées en bas. J'ai vérifié dans nos standards, on n'a rien écrit à ce sujet mais ça vaudrait sûrement le coup pour clarifier

Puisque `_toAuthenticationMethodsDomain` appelle `_getAuthenticationComplementAndExternalIdentifier`, j'aurais aussi inversé leur ordre de lecture.

chrc replied:
Done.
En lecture vers le bas, `_getAuthenticationComplementAndExternalIdentifier` apparait avant `_toAuthenticationMethodsDomain`.

thloesch commented on api/lib/infrastructure/repositories/user-repository.js:
```diff
@@ -109,6 +114,46 @@ function _toPixRolesDomain(pixRolesBookshelf) {
+  let externalIdentifier = authenticationMethodBookshelf.get('externalIdentifier');
+
+  if (identityProvider === AuthenticationMethod.identityProviders.PIX) {
+    authenticationComplement = new AuthenticationMethod.PasswordAuthenticationMethod({
```
**Suggestion Bloquante**

```suggestion
    authenticationComplement = new AuthenticationMethod.PasswordAuthenticationComplement({
```

chrc replied:
Si la class `PasswordAuthenticationMethod` de la class `AuthenticationMethod` doit être renommée, et pour rester cohérent avec `PoleEmploiAuthenticationComplement`, je choisirai `PixAuthenticationComplement`.

chrc replied:
Done

thloesch commented on api/tests/integration/infrastructure/repositories/authentication-method-repository_test.js:
```diff
@@ -200,11 +262,141 @@ describe('Integration | Repository | AuthenticationMethod', () => {
+      await databaseBuilder.commit();
+    });
+
+    it('should create password and set shouldChangePassword to true', async () => {
```
**BSR non-bloquant** : On aurait bien vu un second test pour vérifier que l'utilisateur a désormais 2 authenticationMethods. 

Si la nouvelle méthode écrase la première, ce test ne nous le dit pas

chrc replied:
Il n'y a pas de méthode dans `user-repository` ou `authentication-method-repository` pour récupérer l'ensemble des `AuthenticationMethod` d'un user.

Voulez-vous faire deux appels à `authenticationMethodRepository.findOneByUserIdAndIdentityProvider()`, pour récupérer les `AuthenticationMethod` (`identityProvider.PIX` et `identityProvider.GAR`), et verifier leur contenu ?

chrc replied:
Je fais comme ça.
Done

AndreiaPena commented on api/tests/unit/domain/services/encryption-service_test.js:
```diff
@@ -1,47 +1,45 @@
       // given
-      const encryptedPassword = bcrypt.hashSync('my-real-password', 1);
-      const password = 'my-expected-password';
+      const rawPassword = 'my-expected-password';
+      const hashedPassword = 'ABCDEF1234';
```
**Suggestion Non-bloquante** Pour éviter les ambiguïtés sur le système de hachage, réutiliser ce qui existait avant coté hashSync.
```suggestion 
const rawPassword = 'my-expected-password';
const hashedPassword = bcrypt.hashSync('ABCDEF1234', 1);
```

chrc replied:
On veut comparer `rawPassword` et `hashedPassword` et vérifier le comportement de la fonction `checkPassword()` si `hashedPassword` ne correspond pas effectivement au hash.

Il n'est pas besoin ne connaitre le résultat de `bcrypt.hashSync()`. On peut stuber la fonction, ou plus simplement affecter un hash erroné à `hashedPassword` (ex. `my-expected-password`, ou `wrongHash`).

AndreiaPena commented on api/tests/unit/domain/services/encryption-service_test.js:
```diff
@@ -1,47 +1,45 @@
-      const encryptedPassword = bcrypt.hashSync('my-real-password', 1);
-      const password = 'my-real-password';
+      const rawPassword = 'ABCDEF1234';
+      const hashedPassword = await bcrypt.hashSync(rawPassword, 1);
```
**Remarque** pas besoin du await ici.

chrc replied:
Done

thloesch commented on api/tests/unit/domain/services/user-service_test.js:
```diff
@@ -0,0 +1,159 @@
+
+const { expect, sinon, domainBuilder } = require('../../../test-helper');
+
+const DomainTransaction = require('../../../../lib/infrastructure/DomainTransaction');
```
**Remarque Non-bloquante** On devrait pouvoir stub le DomainTransaction sans avoir à importer le vrai.
Mais on n'a pas réussi à trouver une suggestion pour y remédier.

chrc replied:
Oui c'est un objet complexe que je n'ai pas réussi à remplacer dans ce test unitaire.
Malgré tout, heureux qu'il fonctionne partiellement stubé ici :)
